crates/bevy_ecs/src/entity/mod.rs

```diff
@@ -1202,66 +1152,114 @@ impl Entities {
+    /// It probably had the wrong generation or was created erroneously.
+    #[error("Invalid id: {0}")]
+    InvalidId(EntityDoesNotExistError),
+    /// The [`Entity`] to construct was already constructed.
+    #[error("The entity can not be constructed as it already has a location.")]
+    AlreadyConstructed,
+}
+
 /// An error that occurs when a specified [`Entity`] does not exist.
 #[derive(thiserror::Error, Debug, Clone, Copy, PartialEq, Eq)]
-#[error("The entity with ID {entity} {details}")]
+#[error(
+    "The entity with ID {entity} does not exist; its row now has generation {current_generation}."
+)]
 pub struct EntityDoesNotExistError {
     /// The entity's ID.
     pub entity: Entity,
-    /// Details on why the entity does not exist, if available.
-    pub details: EntityDoesNotExistDetails,
+    /// The generation of the [`EntityRow`], which did not match the requested entity.
+    pub current_generation: EntityGeneration,
 }
 
-impl EntityDoesNotExistError {
-    pub(crate) fn new(entity: Entity, entities: &Entities) -> Self {
-        Self {
-            entity,
-            details: entities.entity_does_not_exist_error_details(entity),
-        }
-    }
-}
-
-/// Helper struct that, when printed, will write the appropriate details
-/// regarding an entity that did not exist.
-#[derive(Copy, Clone, Debug, Eq, Hash, Ord, PartialEq, PartialOrd)]
-pub struct EntityDoesNotExistDetails {
-    location: MaybeLocation<Option<&'static Location<'static>>>,
+/// An error that occurs when a specified [`Entity`] exists but was not constructed when it was expected to be.
+#[derive(thiserror::Error, Debug, Clone, Copy, PartialEq, Eq)]
+pub struct EntityNotConstructedError {
+    /// The entity's ID.
+    pub entity: Entity,
+    /// The location of what last destructed the entity.
+    pub location: MaybeLocation<&'static Location<'static>>,
 }
 
-impl fmt::Display for EntityDoesNotExistDetails {
+impl fmt::Display for EntityNotConstructedError {
     fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        let entity = self.entity;
         match self.location.into_option() {
-            Some(Some(location)) => write!(f, "was despawned by {location}"),
-            Some(None) => write!(
-                f,
-                "does not exist (index has been reused or was never spawned)"
-            ),
+            Some(location) => write!(f, "The entity with ID {entity} is not constructed; its row was last destructed by {location}."),
             None => write!(
                 f,
-                "does not exist (enable `track_location` feature for more details)"
+                "The entity with ID {entity} is not constructed; enable `track_location` feature for more details."
             ),
         }
     }
 }
 
+/// Represents an error of either [`EntityDoesNotExistError`] or [`EntityNotConstructedError`].
+#[derive(thiserror::Error, Copy, Clone, Debug, Eq, PartialEq)]
+pub enum ConstructedEntityDoesNotExistError {
```

Comment:
I don't think this is a good name. "Constructed" entities _always_ exist by definition, "not existing" is only one of the two variants, and the other variant is `ConstructedEntityDoesNotExistError::WasNotContructed`, which is a contradiction.